### PR TITLE
DetailsList: adding an optional column re-ordering feature

### PR DESCRIPTION
**Pull request by laxmikankanala**
#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [ ] Include a change request file using `$ npm run change`

#### Description of changes

Column Reorder support in DetailsList using drag and drop
Based on the mouse position of the event, column target position will be decided.
There is an optional parameter to specify frozen columns to keep the position of those columns intact.

Pending things:
1. Need to test this with all different options/types of DetailsList
2. Need to fix the hit area for drag, that is being impacted by marquee selection
3. Add CSS changes to show the drag items and drop hints

#### Focus areas to test
Functionality of Column reorder

### CODE REVIEW

**Comment by JasonGore** [comment]
This would definitely be a `minor` change since it touches multiple interfaces. 

Please add tests for the new DetailsColumn component and consider making test cases for DetailsHeader and/or DetailsList components to exercise the new functionality. This will help ensure nobody breaks the functionality you've added. 😃

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts [hunk 1]
```diff
@@ -417,6 +423,19 @@ export enum ConstrainMode {
...
 }
 
+export interface IColumnReorderOptions {
+  /** Specifies the number columns with position fixed from left(0th index))
+   * @default 0
+   */
+  frozenColumnCount?: number;
+
+  /** Callback to handle the column reorder
+   * draggaedIndex is the source column index, that need to be placed in targetIndex
```
Comment (JasonGore) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts]: Typo

File: packages/office-ui-fabric-react/src/components/DetailsList/examples/DetailsList.DragDrop.Example.tsx [hunk 2]
```diff
@@ -1,57 +1,89 @@
...
-            setKey='items'
-            items={ items }
-            selection={ this._selection }
-            selectionPreservedOnEmptyClick={ true }
-            onItemInvoked={ this._onItemInvoked }
-            onRenderItemColumn={ this._onRenderItemColumn }
-            dragDropEvents={ this._getDragDropEvents() }
-          />
-        </MarqueeSelection>
+        {/* <MarqueeSelection selection={ this._selection }> */ }
```
Comment (JasonGore) [inline packages/office-ui-fabric-react/src/components/DetailsList/examples/DetailsList.DragDrop.Example.tsx]: Why was `MarqueeSelection` removed? Is there a reason to leave it commented in?

**Comment by shagra-ms** [comment]
@amitdh is added to the review. #Closed

File: packages/office-ui-fabric-react/src/utilities/dragdrop/interfaces.ts [hunk 3]
```diff
@@ -18,6 +18,7 @@ export interface IDragDropEvents {
   onDrop?: (item?: any, event?: DragEvent) => void;
   onDragStart?: (item?: any, itemIndex?: number, selectedItems?: any[], event?: MouseEvent) => void;
   onDragEnd?: (item?: any, event?: DragEvent) => void;
+  onDragOver?: (item?: any, event?: DragEvent) => void;
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/utilities/dragdrop/interfaces.ts]: >onDragOver?: (item?: any, event?: DragEvent) => void; [](start = 2, length = 53)

Do we have to expose this in IDragDropEvents interface?

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.scss [hunk 4]
```diff
@@ -123,6 +123,21 @@ $isPaddedMargin: 24px;
...
+  top: 100;
+  overflow: hidden;
+  height: inherit;
+  width: 2px;
+  z-index:100;
+
+  &.isDropping {
+    background: rgb(82, 79, 79);
+  }
+}
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.scss]: Is this according to the UX? Or is this just a placeholder for now?

Comment (laxmikankanala) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.scss]: For now, just a place holder to see it getting updated while dragging.

File: packages/office-ui-fabric-react/src/utilities/dragdrop/interfaces.ts (hunk 3 above)
Comment (laxmikankanala) [reply packages/office-ui-fabric-react/src/utilities/dragdrop/interfaces.ts]: Yes, as this is the only event that gets triggered while dragging on a cell, and the drop hint status needs to be changed while crossing the 50% of cell width

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx [hunk 5]
```diff
@@ -218,96 +259,113 @@ export class DetailsHeader extends BaseComponent<IDetailsHeaderProps, IDetailsHe
...
+                  </div>,
+                  column.ariaLabel && !this.props.onRenderColumnHeaderTooltip ? (
+                    <label
+                      key={ `${column.key}_label` }
+                      id={ `${this._id}-${column.key}-tooltip` }
+                      className={ styles.accessibleLabel }
+                    >
+                      { column.ariaLabel }
+                    </label>
+                  ) : null,*/
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx]: >) : null,*/ [](start = 18, length = 11)

Please remove commented code from here.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx [hunk 6]
```diff
@@ -218,96 +259,113 @@ export class DetailsHeader extends BaseComponent<IDetailsHeaderProps, IDetailsHe
...
-                          aria-labelledby={ column.isIconOnly ? undefined : `${this._id}-${column.key}-name ` }
-                          className={ css('ms-DetailsHeader-cellTitle', styles.cellTitle) }
-                          data-is-focusable={ column.columnActionsMode !== ColumnActionsMode.disabled }
-                          role={ column.columnActionsMode !== ColumnActionsMode.disabled ? 'button' : undefined }
-                          aria-describedby={ this.props.onRenderColumnHeaderTooltip ? `${this._id}-${column.key}-tooltip` : undefined }
-                          onContextMenu={ this._onColumnContextMenu.bind(this, column) }
-                          onClick={ this._onColumnClick.bind(this, column) }
-                          aria-haspopup={ column.columnActionsMode === ColumnActionsMode.hasDropdown }
-                        >
+                (columnReorderOptions && _isDraggable) && this._renderDropHint(columnIndex),
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx]: >columnReorderOptions && [](start = 17, length = 23)

Required?

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx [hunk 7]
```diff
@@ -218,96 +259,113 @@ export class DetailsHeader extends BaseComponent<IDetailsHeaderProps, IDetailsHe
...
+                    >
+                      { column.ariaLabel }
+                    </label>
+                  ) : null,*/
+                ),
+                (column.isResizable) && this._renderColumnSizer(columnIndex),
               ]
             );
-          })
+          }) }
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx]: >} [](start = 13, length = 1)

Please fix indentation. This should go to next line.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx [hunk 8]
```diff
@@ -326,7 +384,131 @@ export class DetailsHeader extends BaseComponent<IDetailsHeaderProps, IDetailsHe
...
+      onDragOver: this._dragDropColumnEvents!.onDragOver
+    };
+    return options;
+  }
+
+  private _updateDroppingState(newValue: boolean, event: DragEvent): void {
+    // TODO - Handle CSS changes
+  }
+
+  private _getDragDropColumnEvents(): IDragDropEvents {
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx]: >private _getDragDropColumnEvents(): IDragDropEvents [](start = 2, length = 51)

Why is this required? We can do whatever we are doing here inside IDragDropOptions.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.scss (hunk 4 above)
Comment (laxmikankanala) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.scss]: For now, just a placeholder

---
In reply to: [188871768](https://github.com/OfficeDev/office-ui-fabric-react/pull/4857#discussion_r188871768) [](ancestors = 188871768)

File: packages/office-ui-fabric-react/src/utilities/dragdrop/DragDropHelper.tsx [hunk 9]
```diff
@@ -369,6 +372,8 @@ export class DragDropHelper implements IDragDropHelper {
           this._events.on(activeTarget.target.root, 'mouseleave', this._onMouseLeave.bind(this, activeTarget.target));
         }
       }
+      // To prevent marquee selection on draggable targets
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/utilities/dragdrop/DragDropHelper.tsx]: >To prevent [](start = 9, length = 10)

Or simply - 
'Prevent marquee selection on draggable targets'

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+    this._root = createRef();
+    this._onDragStart = this._onDragStart.bind(this);
+    this._onDragEnd = this._onDragEnd.bind(this);
+  }
+
+  public render() {
+    const { column, columnIndex, parentId } = this.props;
+    const { onRenderColumnHeaderTooltip = this._onRenderColumnHeaderTooltip
+    } = this.props;
+    const isDraggable = this.props.isDraggable;
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx]: >const isDraggable = this.props.isDraggable; [](start = 4, length = 43)

merge this with 2 above ?

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+
+  private _onDragStart(item?: any, itemIndex?: number, selectedItems?: any[], event?: MouseEvent): void {
+    if (itemIndex) {
+      this.props.setDraggedItemIndex!(itemIndex);
+    }
+  }
+
+  private _onDragEnd(item?: any, event?: MouseEvent): void {
+    this.props.setDraggedItemIndex!(-1);
+    event!.preventDefault();
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx]: >event!.preventDefault(); [](start = 4, length = 24)

Why is this required?

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx [hunk 12]
```diff
@@ -326,7 +303,129 @@ export class DetailsHeader extends BaseComponent<IDetailsHeaderProps, IDetailsHe
...
+      onDragEnd: () => undefined,
+      onDragOver: this._onDragOver
+    };
+    return options;
+  }
+
+  private _updateDroppingState(newValue: boolean, event: DragEvent): void {
+    if ((newValue === false || this._draggedColumnIndex === -1) && event.type !== 'drop') {
+      const newDropHintState = this.state.dropHintsState!.map(state => false);
+      this.setState({ dropHintsState: newDropHintState });
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx]: >this.setState({ dropHintsState: newDropHintState }); [](start = 6, length = 52)

Pass in a function instead. This way of updating the state is not reliable.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+  componentRef?: () => void;
+  column: IColumn;
+  columnIndex: number;
+  parentId?: string;
+  onRenderColumnHeaderTooltip?: IRenderFunction<ITooltipHostProps>;
+  onColumnClick?: (ev: React.MouseEvent<HTMLElement>, column: IColumn) => void;
+  onColumnContextMenu?: (column: IColumn, ev: React.MouseEvent<HTMLElement>) => void;
+  dragDropHelper?: IDragDropHelper | null;
+  isDraggable?: boolean;
+  setDraggedItemIndex?: (itemIndex: number) => void;
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx]: >setDraggedItemIndex?: (itemIndex: number) => void; [](start = 2, length = 50)

Rather than this, the notification should be simple - like 'onColumnDragStart' with no params. The params (index) can be handled inside the DetailsHeader, where this callback is binded to a function.

**Comment by betrue-final-final** [comment]
Hey @dzearing and @aditima, can you peeps check this one out?

**Comment by laxmikankanala** [comment]
@erichdev, the issue of frequent renders has been fixed in the latest commit. Now DetailsHeader will get re-rendered only when there is a change in the drop hints.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx [hunk 14]
```diff
@@ -121,16 +163,30 @@ export class DetailsHeader extends BaseComponent<IDetailsHeaderProps, IDetailsHe
...
+    const { columns, ariaLabel, ariaLabelForSelectAllCheckbox, selectAllVisibility, ariaLabelForSelectionColumn, columnReorderOptions, onColumnClick, onColumnContextMenu } = this.props;
     const { isAllSelected, columnResizeDetails, isSizing, groupNestingDepth, isAllCollapsed } = this.state;
-
     const showCheckbox = selectAllVisibility !== SelectAllVisibility.none;
-
     const {
       onRenderColumnHeaderTooltip = this._onRenderColumnHeaderTooltip
     } = this.props;
 
+    if (!this._dragDropHelper) {
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx]: >if (!this._dragDropHelper) [](start = 4, length = 26)

if (!this._dragDropHelper && this.props.columenReorderOptions) {
this._dragDropHelper = new DragDropHelper();
}

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx [hunk 15]
```diff
@@ -329,6 +319,113 @@ export class DetailsHeader extends BaseComponent<IDetailsHeaderProps, IDetailsHe
...
+
+  private _setDraggedItemIndex(itemIndex: number) {
+    if (itemIndex >= 0) {
+      this._draggedColumnIndex = (this.props.selectionMode !== SelectionMode.none) ? itemIndex! - 1 : itemIndex;
+      this._getDropHintPositions();
+    } else {
+      this._draggedColumnIndex = -1;
+    }
+
+  }
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx]: Remove the extra line.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx [hunk 16]
```diff
@@ -329,6 +319,113 @@ export class DetailsHeader extends BaseComponent<IDetailsHeaderProps, IDetailsHe
...
+
+  private _onDragOver(item: any, event: DragEvent): void {
+    if (this._draggedColumnIndex !== -1) {
+      const clientX = event.clientX;
+      this._updateDropHintStates(clientX);
+    }
+  }
+
+  private _onDrop(item?: any, event?: DragEvent): void {
+    if (this._draggedColumnIndex !== -1 && event! instanceof DragEvent) {
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx]: >this._draggedColumnIndex !== -1 [](start = 8, length = 31)

A better check for 'valid' draggedColumnIndex would be a non-negative check.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx [hunk 17]
```diff
@@ -329,6 +319,113 @@ export class DetailsHeader extends BaseComponent<IDetailsHeaderProps, IDetailsHe
...
+   *
+   */
+  private _updateDropHintStates = (eventXposition: number): void => {
+    let currentIndex: number = (this.props.columnReorderOptions!.frozenColumnCount) ? this.props.columnReorderOptions!.frozenColumnCount! : 0;
+    const colCount = this.props.columns.length;
+    const { currentDropHintIndex } = this.state;
+    let indexToUpdate = -1;
+    if (eventXposition <= this._dropHintOriginXValues[currentIndex]) {
+
+      indexToUpdate = currentIndex;
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx]: NIT - Extra line.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx [hunk 18]
```diff
@@ -329,6 +319,113 @@ export class DetailsHeader extends BaseComponent<IDetailsHeaderProps, IDetailsHe
...
+  }
+
+  private _onDragOver(item: any, event: DragEvent): void {
+    if (this._draggedColumnIndex !== -1) {
+      const clientX = event.clientX;
+      this._updateDropHintStates(clientX);
+    }
+  }
+
+  private _onDrop(item?: any, event?: DragEvent): void {
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx]: >_onDrop [](start = 10, length = 7)

We should also reset - _dropHintOriginXValues

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+}
+
+export interface IDetailsColumnState {
+  isDropping: boolean;
+}
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx]: Not used? Please remove.

**Comment by betrue-final-final** [comment]
Alright, I have it built, but how do you see this behavior? I'm not seeing a sample and I can't drag or drop the columns.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.scss [hunk 20]
```diff
@@ -17,7 +17,7 @@ $detailsList-text-color: $listTextColor;
 }
 
 .rootIsHorizontalConstrained {
-  overflow-x: auto;
+  overflow-x: visible;
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.scss]: >-x: visible; [](start = 10, length = 12)

Please ensure there is no regression with this change.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx [hunk 21]
```diff
@@ -83,34 +88,95 @@ export class DetailsHeader extends BaseComponent<IDetailsHeaderProps, IDetailsHe
...
+  private _dragDropHelper: DragDropHelper | null;
+  private _subscriptionObject:
+    {
+      key: string;
+      dispose(): void;
+    };
+  private _onDropIndexInfo: {
+    sourceIndex: number;
+    targetIndex: number;
+  };
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx]: New line - before and after.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx [hunk 22]
```diff
@@ -121,16 +187,30 @@ export class DetailsHeader extends BaseComponent<IDetailsHeaderProps, IDetailsHe
...
     const { isAllSelected, columnResizeDetails, isSizing, groupNestingDepth, isAllCollapsed } = this.state;
-
     const showCheckbox = selectAllVisibility !== SelectAllVisibility.none;
-
     const {
       onRenderColumnHeaderTooltip = this._onRenderColumnHeaderTooltip
     } = this.props;
 
+    if (!this._dragDropHelper) {
+      this._dragDropHelper = this.props.columnReorderOptions ? new DragDropHelper({
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx]: bad merge? Please revert.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx [hunk 23]
```diff
@@ -218,97 +298,32 @@ export class DetailsHeader extends BaseComponent<IDetailsHeaderProps, IDetailsHe
...
+                    column={ column }
+                    key={ column.key }
+                    columnIndex={ (showCheckbox ? 1 : 0) + columnIndex }
+                    parentId={ this._id }
+                    isDraggable={ _isDraggable }
+                    setDraggedItemIndex={ this._setDraggedItemIndex }
+                    dragDropHelper={ this._dragDropHelper }
+                    onColumnClick={ onColumnClick }
+                    onColumnContextMenu={ onColumnContextMenu }
+                    isDropped={ this._onDropIndexInfo.targetIndex === columnIndex + 1 }
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx]: >columnIndex + 1 [](start = 70, length = 15)

Is showCheckbox check required?

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx [hunk 24]
```diff
@@ -329,6 +344,115 @@ export class DetailsHeader extends BaseComponent<IDetailsHeaderProps, IDetailsHe
...
+    }
+    // TODO - Handle CSS changes
+  }
+
+  private _isValidCurrentDropHintIndex() {
+    return this.state.currentDropHintIndex! >= 0;
+  }
+
+  private _onDragOver(item: any, event: DragEvent): void {
+    if (this._draggedColumnIndex !== -1) {
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx]: >!== -1 [](start = 33, length = 6)

Bad merge. Please revert.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx [hunk 25]
```diff
@@ -329,6 +344,115 @@ export class DetailsHeader extends BaseComponent<IDetailsHeaderProps, IDetailsHe
...
+      const clientX = event.clientX;
+      this._updateDropHintStates(clientX);
+    }
+  }
+
+  private _onDrop(item?: any, event?: DragEvent): void {
+    if (this._draggedColumnIndex !== -1 && event! instanceof DragEvent) {
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx]: >!== -1 [](start = 33, length = 6)

Bad merge. Please revert.

**Comment by shagra-ms** [comment]
>       this._dropHintOriginXValues = {};

Bad merge. Please revert.

---
Refers to: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx:372 in 28e5399. [](commit_id = 28e53991c8c3d04a5370aac4ae9658a6316fb4db, deletion_comment = True)

**Comment by shagra-ms** [comment]
>     this._dropHintOriginXValues = {};

Bad merge. Please revert.

---
Refers to: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx:362 in 28e5399. [](commit_id = 28e53991c8c3d04a5370aac4ae9658a6316fb4db, deletion_comment = True)

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx [hunk 26]
```diff
@@ -0,0 +1,264 @@
...
+export class DetailsColumn extends BaseComponent<IDetailsColumnProps, IDetailsColumnState> {
+  private _root: any;
+  private _dragDropSubscription: IDisposable;
+
+  constructor(props: IDetailsColumnProps) {
+    super(props);
+
+    this.state = {
+      isDropping: false,
+    };
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx]: Bad merge. Please revert.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx [hunk 27]
```diff
@@ -0,0 +1,264 @@
...
+  onColumnContextMenu?: (column: IColumn, ev: React.MouseEvent<HTMLElement>) => void;
+  dragDropHelper?: IDragDropHelper | null;
+  isDraggable?: boolean;
+  setDraggedItemIndex?: (itemIndex: number) => void;
+  isDropped?: boolean;
+}
+
+export interface IDetailsColumnState {
+  isDropping: boolean;
+}
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx]: Bad merge. Please revert.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx [hunk 28]
```diff
@@ -83,34 +88,95 @@ export class DetailsHeader extends BaseComponent<IDetailsHeaderProps, IDetailsHe
...
+  }
+
+  public componentWillUpdate(nextProps: IDetailsHeaderProps, nextState: IDetailsHeaderState): void {
+    if (this.props !== nextProps && this._onDropIndexInfo.sourceIndex >= 0 && this._onDropIndexInfo.targetIndex >= 0) {
+      if (this.props.columns[this._onDropIndexInfo.sourceIndex] === nextProps.columns[this._onDropIndexInfo.sourceIndex] && this.props.columns[this._onDropIndexInfo.targetIndex] === nextProps.columns[this._onDropIndexInfo.targetIndex]) {
+        this._onDropIndexInfo = {
+          sourceIndex: Number.MIN_SAFE_INTEGER,
+          targetIndex: Number.MIN_SAFE_INTEGER
+        };
+      }
```
Comment (shagra-ms) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.tsx]: I would actually prefer a different condition. Before we re-render, it is important to confirm that onDrop changes has actually been applied. A check for that would be -
this.props.columns[sourceIndex] === newProps.columns[targetIndex]
And if this is not true, we should discard the onDropIndexInfo.

File: common/changes/office-ui-fabric-react/laxmika-colreorder_2018-05-13-07-07.json [hunk 29]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "office-ui-fabric-react",
+      "comment": "Added column reorder with DragAndDrop support in Details List. As part of this feature, a new component DetailsColumn has been added inside the header, for each  column.  An optional new prop(ColumnReorderOptions) has been added to DetailsList to handle the column reorder. Have added column level drag subscriptions to be able to drag the columns, and added one header level subscription, to handle the drops. All the drag drop events are being handled at header level.",
+      "type": "patch"
```
Comment (yiminwu) [inline common/changes/office-ui-fabric-react/laxmika-colreorder_2018-05-13-07-07.json]: >patch [](start = 15, length = 5)

This should be 'minor'

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx [hunk 30]
```diff
@@ -0,0 +1,255 @@
...
+    if (this.props.setDraggedItemIndex) {
+      this.props.setDraggedItemIndex(-1);
+      this._root.current.classList.remove(styles.borderWhileDragging);
+    }
+  }
+
+  private _onColumnContextMenu(column: IColumn, ev: React.MouseEvent<HTMLElement>): void {
+    const { onColumnContextMenu } = this.props;
+    if (column.onColumnContextMenu) {
+      column.onColumnContextMenu(column, ev);
```
Comment (aditima) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx]: Just curious, why we need both column.onColumnContextMenu and props.onColumnContextMenu. We'll end up executing both if they are provided, not sure if that is intentional too.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.scss [hunk 31]
```diff
@@ -220,6 +220,7 @@ $detailsList-item-focus-meta-text-color: $ms-color-neutralDark;
   white-space: nowrap;
   overflow: hidden;
   text-overflow: ellipsis;
+  padding-left:12px;
```
Comment (aditima) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsRow.scss]: >padding-left [](start = 2, length = 12)

Use @include padding-left(12px) instead for RTL support

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx [hunk 32]
```diff
@@ -0,0 +1,255 @@
...
+          'ms-DetailsHeader-cell',
+          styles.cell,
+          column.headerClassName,
+          column.columnActionsMode !== ColumnActionsMode.disabled && 'is-actionable ' + styles.cellIsActionable,
+          !column.name && 'is-empty ' + styles.cellIsEmpty,
+          (column.isSorted || column.isGrouped || column.isFiltered) && 'is-icon-visible',
+          column.isPadded && styles.cellWrapperPadded
+        )}
+        data-is-draggable={isDraggable}
+        draggable={isDraggable}
```
Comment (aditima) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx]: Why do we need both data-is-draggable and draggable attributes?

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx (hunk 30 above)
Comment (laxmikankanala) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx]: This is an existing function which was there in ColumnHeader, i just moved it as is.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx (hunk 32 above)
Comment (laxmikankanala) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx]: I have followed DetailsRow, may be kept both for legacy reasons,

**Comment by betrue-final-final** [comment]
Hey @laxmikankanala, it looks like the values are truncated way too early. That's why I rejected the screener test. Can you check that out?

**Comment by shagra-ms** [comment]
Hey Ben,

I just looked at the screenshots for screener test. Seems that these are expected after the increase in left padding (+5px). This regression was called out during the design iteration as well. @Piyush<mailto:Piyush.Belchandan@microsoft.com> @Niket<mailto:Jain.Niket@microsoft.com>

~Shreyansh Agrawal

________________________________

**Comment by betrue-final-final** [comment]
I did understand that, but it looks like it's still truncating way too early.
image

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx (hunk 30 above)
Comment (aditima) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx]: Ah I see... thanks!

---
In reply to: [196748992](https://github.com/OfficeDev/office-ui-fabric-react/pull/4857#discussion_r196748992) [](ancestors = 196748992)

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx (hunk 32 above)
Comment (aditima) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx]: @yiminwu @ThomasMichon - do any of you recall why we need both these attributes? Seems sort of redundant for the new component.

Comment (yiminwu) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsColumn.tsx]: I didn't remember.

---
In reply to: [196860743](https://github.com/OfficeDev/office-ui-fabric-react/pull/4857#discussion_r196860743) [](ancestors = 196860743)

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.scss [hunk 33]
```diff
@@ -75,6 +75,18 @@ $isPaddedMargin: 24px;
   &.cellIsEmpty {
     text-overflow: clip;
   }
+
+  &:hover .GripperBarVerticalStyle {
+    display: block;
+   }
+}
+
+.GripperBarVerticalStyle {
```
Comment (yiminwu) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.scss]: nit: can we change to start with lowcase to match others?

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.scss [hunk 34]
```diff
@@ -75,6 +75,18 @@ $isPaddedMargin: 24px;
   &.cellIsEmpty {
     text-overflow: clip;
   }
+
+  &:hover .GripperBarVerticalStyle {
```
Comment (yiminwu) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.scss]: >GripperBarVerticalStyle [](start = 11, length = 23)

same here

**Comment by laxmikankanala** [comment]
@betrue-final-final, we have fixed the right padding issue. Can you please review now.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.scss (hunk 34 above)
Comment (laxmikankanala) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.scss]: Fixed

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.scss (hunk 33 above)
Comment (laxmikankanala) [reply packages/office-ui-fabric-react/src/components/DetailsList/DetailsHeader.scss]: Fixed

**Comment by betrue-final-final** [comment]
I wasn't quite up then, but I just checked it. It was 5:30AM PST ✌️
